### DISCUSSION TITLE
Workarounds for LSP server download failure

### DISCUSSION THREAD

**Discussion by andrew-ares** [zed-industries/zed#6594] [Show and tell]
You may encounter LSP server download failures if you are using Zed in China, as seen in issue #1352, #1377. Here are two possible workarounds to address this error:

1. (Recommended) Search for the “Copy shell command” related option or setting in your VPN software. For instance, [clashX](https://github.com/yichengchen/clashX) offers the following script:

```bash
export https_proxy=http://127.0.0.1:7890 http_proxy=http://127.0.0.1:7890 all_proxy=socks5://127.0.0.1:7890
```
Execute the script and run `zed <your project path>` in the terminal. This should work now.

*However*, there are two things to keep in mind:

a. Due to the instability of the Great Firewall (GFW), the LSP server may sometimes download slowly or fail to download altogether, depending on the GFW's mood :( . If this happens, temporary download files will be left in the directory `$HOME/Library/Application Support/Zed/languages` when you exit Zed. It's important to delete these files to avoid clutter. For instance, in the case of Rust, we should delete the `rust-analyzer` folder.

b. Since Zed automatically downloads the latest LSP server at startup, It's recommended to launch Zed from a shell where the `http_proxy` env variable is set. This step will no longer be necessary once issue #887 is resolved. 

2. From the above, we know that we can manually download the LSP server. Let's take Rust as an example:

```shell
$ cd "$HOME/Library/Application Support/Zed/languages"
$ mkdir rust-analyzer # if not exists
$ cd rust-analyzer
$ wget https://github.com/rust-lang/rust-analyzer/releases/download/2023-04-04/rust-analyzer-aarch64-apple-darwin.gz 
$ gzip -d rust-analyzer-aarch64-apple-darwin.gz # we get rust-analyzer-aarch64-apple-darwin
$ mv rust-analyzer-aarch64-apple-darwin rust-analyzer-2023-04-04 # depends on the release date
$ chmod +x rust-analyzer-2023-04-04
```
This method can be useful when there is no network.

**Comment by lifesign** [comment]
thx bro~
